Content.Client/Administration/UI/BwoinkWindow.xaml.cs

```diff
@@ -17,38 +15,65 @@ namespace Content.Client.Administration.UI
     {
-        [Dependency] private readonly IEntitySystemManager _systemManager = default!;
-        [Dependency] private readonly IPlayerManager _playerManager = default!;
-
-        private readonly NetUserId _channelId;
-
-        public BwoinkWindow(NetUserId userId, string channelName)
+        private readonly BwoinkSystem _bwoinkSystem;
+        public BwoinkWindow(BwoinkSystem bs)
         {
             RobustXamlLoader.Load(this);
-            IoCManager.InjectDependencies(this);
+            _bwoinkSystem = bs;
 
-            _channelId = userId;
-            Title = (_playerManager.LocalPlayer?.UserId == _channelId) ? "Admin Message" : channelName;
+            ChannelSelector.OnSelectionChanged += sel =>
+            {
+                if (sel is not null)
+                {
+                    SwitchToChannel(sel.SessionId);
+                    Title = $"{sel.CharacterName} / {sel.Username}";
+                }
 
-            SenderLineEdit.OnTextEntered += Input_OnTextEntered;
+                foreach (var li in ChannelSelector.PlayerItemList)
+                    li.Text = FormatTabTitle(li);
+            };
+
+            ChannelSelector.DecoratePlayer += (PlayerInfo pl, ItemList.Item li) =>
+            {
+                li.Text = FormatTabTitle(li, pl);
+            };
         }
 
-        private void Input_OnTextEntered(LineEdit.LineEditEventArgs args)
+        private string FormatTabTitle(ItemList.Item li, PlayerInfo? pl = default)
         {
-            if (!string.IsNullOrWhiteSpace(args.Text))
+            pl ??= (PlayerInfo) li.Metadata!;
+            var sb = new StringBuilder();
+            sb.Append(pl.Connected ? '●' : '○');
+            sb.Append(' ');
+            if (_bwoinkSystem.TryGetChannel(pl.SessionId, out var panel) && panel.Unread > 0)
             {
-                var bwoink = _systemManager.GetEntitySystem<BwoinkSystem>();
-                bwoink.Send(_channelId, args.Text);
+                if (panel.Unread < 11)
+                    sb.Append(new Rune('➀' + (panel.Unread-1)));
+                else
+                    sb.Append(new Rune(0x2639)); // ☹
+                sb.Append(' ');
             }
 
-            SenderLineEdit.Clear();
+            if (pl.Antag)
+                sb.Append(new Rune(0x1F5E1)); // 🗡
+
+            sb.AppendFormat("\"{0}\"", pl.CharacterName)
+                .Append(' ')
+                .Append(pl.Username);
+
+            return sb.ToString();
         }
 
-        public void ReceiveLine(string text)
+        public void SwitchToChannel(NetUserId ch)
         {
-            var formatted = new FormattedMessage(1);
-            formatted.AddMarkup(text);
-            TextOutput.AddMessage(formatted);
+            foreach (var bw in BwoinkArea.Children)
+                bw.Visible = false;
+
+            var np = _bwoinkSystem.EnsurePanel(ch);
+            if (!BwoinkArea.Children.Contains(np))
+                BwoinkArea.AddChild(np);
```

Comment:
imo ensureadmin should parent it, not this

Reply:
imo `BwoinkSystem` should delegate to the window for UI concerns like this, hence the behavior

Reply:
yet bwoinksystem still creates the ui, opens it, etc. ensuring a panel should always have it be parented to the bwoinkwindow, delegating this to the control and not doing it in the method itself not only invites inproper usage of the method, but also makes the code less readable, since panel creation logic is strewn about multiple places

Reply:
Fixed by 21e7dc8f1de38a916845a7e430cb6f1a9a38debd